p2panda-discovery-next/src/random_walk.rs

```diff
@@ -0,0 +1,383 @@
+// SPDX-License-Identifier: MIT OR Apache-2.0
+
+use std::collections::HashSet;
+use std::hash::Hash as StdHash;
+use std::marker::PhantomData;
+
+use rand::Rng;
+use rand::seq::IteratorRandom;
+use thiserror::Error;
+use tokio::sync::{Mutex, RwLock};
+
+use crate::address_book::{AddressBookStore, NodeInfo};
+use crate::{DiscoveryResult, DiscoveryStrategy};
+
+#[derive(Debug)]
+pub struct RandomWalkerConfig {
+    /// Probability of resetting the random walk and starting from scratch, determined on every
+    /// walking step.
+    ///
+    /// ```text
+    /// 0.0 = Never reset
+    /// 1.0 = Always reset
+    /// ```
+    ///
+    /// Defaults to 0.02 (2%) probability.
+    pub reset_walk_probability: f64,
+}
+
+impl Default for RandomWalkerConfig {
+    fn default() -> Self {
+        Self {
+            reset_walk_probability: 0.02, // 2% chance
+        }
+    }
+}
+
+/// Breadth-first "random walk" peer sampling strategy.
+///
+/// Strategy to traverse a network of nodes of possibly unknown size and shape combined with a
+/// discovery protocol to exchange known nodes.
+///
+/// A random next node is selected to initiate the discovery protocol, starting from a locally
+/// configured set of "bootstrap nodes" if available. The "walker" continues with another random
```

Comment:
Could be good to explain what differentiates a "bootstrap node" from a "normal node" and clarify if both are included in the random walk algorithm.

Reply:
Somehow have a hard time explaining "bootstrap nodes" different than in this sentence right now (they are prioritized on the initial first walk), maybe you have an idea how to frame this better?